### DISCUSSION TITLE
Similar object and background color, how to change background color to white?

### DISCUSSION THREAD

**Discussion by glslmn** [ImageMagick/ImageMagick#1937] [General]
Hi, 

So I have this photo at hand: 
![im9](https://user-images.githubusercontent.com/40963477/80291349-76017b80-8755-11ea-81d8-bee8ec22a44c.jpg)
 and I want to change the background color to white (so I want anything around the phone object to be white). To do so, I use the following code:
```
convert source_image -fuzz XX%% -fill white -draw "color 0,0 floodfill" target_image
```
my best output with the optimal XX%% choice is:![im10](https://user-images.githubusercontent.com/40963477/80291429-2ec7ba80-8756-11ea-9a04-f0ef9aa8216b.png)
I know that this happens because both the color of the phone/phone's screen and background colors are very very similar and dark. Is there any way that I can achieve what I want with ImageMagick?

Thank you!

**Comment by fmw42** [comment]
Try this, provided you have a current enough version of ImageMagick 6 to support the new connected components defines:

```
convert img.jpg \
\( -clone 0 -fuzz 15% -fill none -draw "matte 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none \
-alpha off -morphology erode diamond:1 -type bilevel \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=1 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
img_result.png

```

![img_result](https://user-images.githubusercontent.com/13458218/80291905-8a6a5780-8706-11ea-97ed-858191bd61b2.png)

This may only work on this image.

**Reply by glslmn** [reply]
I have ImageMagick 7.0.10-7 Q16 x86_64 2020-04-21 installed. When I run your code, I get the image:
![hola](https://user-images.githubusercontent.com/40963477/80292291-502ca480-875e-11ea-9934-df5925032011.jpg)

**Reply by fmw42** [reply]
Change convert to magick and matte to alpha. Does that work?

**Reply by glslmn** [reply]
No, unfortunately it didn't. The output is the same as the above

**Comment by snibgo** [comment]
@glslmn : what is your platform? If Windows BAT, don't forget to double the %%.

**Reply by glslmn** [reply]
I'm on ubuntu. I tried both with % and %%. the result is the same

**Reply by fmw42** [reply]
Post your modified command line.

**Reply by glslmn** [reply]
```
magick source_image \
\( -clone 0 -fuzz XX% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none \
-alpha off -morphology erode diamond:1 -type bilevel \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=1 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
target_image
```

this code outputs :
![img_result](https://user-images.githubusercontent.com/40963477/80318399-86c3f700-8812-11ea-8600-57c57c0f3fba.png)
(sorry that I misinformed you in my previous command). So all I need to do now is filling that transparent part with white :)

**Reply by glslmn** [reply]
so on my machine when I click on the image it is the phone image on the middle and around the image is covered with transparent layer

**Reply by fmw42** [reply]
Correct. I thought you wanted it transparent. If you want it white (or any other color), then add `-background white -compose over -flatten` before the output

```
magick source_image.jpg \
\( -clone 0 -fuzz 15% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none \
-alpha off -morphology erode diamond:1 -type bilevel \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=1 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
-background white -compose over -flatten \
target_image.png

```
In your command, you did not change -fuzz XX% to some value, like 15%.

![target_image](https://user-images.githubusercontent.com/13458218/80318875-fc53a700-87c1-11ea-82f3-7456ae1bc037.png)

**Reply by glslmn** [reply]
Yes, I changed it and 15%% worked very well for this particular image.Thank you very much! Actually, that piece of code works good also for some other images with challenging backgrounds. After turning the background to white, I trim the white background with the code 
```
convert source_image -fuzz XX%% -trim target_image
``` 
is it possible to integrate the above piece of code into the one that you shared?

Also, perhaps more importantly, where do you suggest me to read to learn ImageMagick better?

**Reply by fmw42** [reply]
If you want to trim the result, it would be better to do so at the end:

```
magick source_image.jpg \
\( -clone 0 -fuzz 15% -fill none -draw "alpha 0,0 floodfill" \
-background black -alpha background \
-fill white +opaque none \
-alpha off -morphology erode diamond:1 -type bilevel \
-define connected-components:mean-color=true \
-define connected-components:exclude-header=true \
-define connected-components:keep-top=1 \
-connected-components 8 \) \
-alpha off -compose copy_opacity -composite \
-background white -compose over -flatten \
-trim +repage \
target_image.png
```

For new users, the basic references that I have are:

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdown
https://imagemagick.org/script/porting.php#cli

**Reply by glslmn** [reply]
thank you!

**Comment by snibgo** [comment]
I suggest also:

https://www.imagemagick.org/script/command-line-options.php

**Reply by glslmn** [reply]
thank you!
